text/3064-bitfield.md

```diff
@@ -0,0 +1,401 @@
+- For each field in the C struct whose name is `_`, delete the field name.
+
+The `&` and `&mut` operators cannot be applied to bit-fields.
+
+When a bit-field field is accessed, the abstract machine may also access
+adjacent bit-field fields but not fields that are separated from the field by a
+StructBodyElement that is not a bit-field field. (Note: This paragraph restricts
+the kinds of loads and stores the compiler can perform when accessing a
+bit-field. This paragraph does not need to be specially advertised to users as
+the inability to take references to bit-field fields makes it impossible to
+access adjacent bit-field fields in otherwise sound code.)
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- Bit-fields cannot be assigned without overflow checks.
+- The annotation feels somewhat far away from the type itself.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+- Alternative: Using `u32: N` to specify the width of a bit-field. I've chosen
+  to use an annotation instead because it seems more likely to be accepted.
+
+- Alternative: Using a dedicated type instead of an annotation:
+
+    ```rust
+    #[repr(C)]
+    struct X {
+        a: BitField<u32, 6>,
+    }
+    ```
+
+    where `BitField` would be some kind of lang item that ensures that
+
+    - the address of a field with such a type cannot be taken
+    - the type and width parameters are correct.
+
+    This type would then have the following API:
+
+    ```rust
+    impl<T, const N: usize> BitField<T, N> {
+        fn new(t: T) -> Self;
+        fn wrapping_new(t: T) -> Self;
+        fn get(self) -> T;
+        fn set(&mut self, t: T);
+        fn wrapping_add(self, t: T) -> Self;
+        ...
+    }
+
+    impl<T, const N: usize> AddAssign<T> for BitField<T, N> {
+    }
+    ```
+
+    The layout of `BitField<T, N>` would be the layout of `T` when used outside
+    a struct.
+
+    This could then be used like so:
+
+    ```rust
+    fn f(x: X) {
+        x.a.set(1);
+        x.a += 1;
+        x.a = x.a.wrapping_add(12);
+        println!("{}", x.a.get());
+    }
+    ```
+
+    There is some magic happening here:
+
+    - `x.a += 1` cannot be written as `AddAssign::add(&mut x.a, 1)` because the
+      address of `x.a` cannot be taken. The compiler would figure it out when
+      the lhs is a struct field.
+    - When the assignment happens, the compiler writes the bits on the rhs to
+      the correct positions in the struct.
+
+    This is a more structured solution but would need to be fleshed out.
+
+# Prior art
+[prior-art]: #prior-art
+
+- C
+- The author of Zig [proposes][zig-proposes] a different syntax that basically
+  boils down to arbitrarily sized integer types:
+
+  ```rust
+  struct X {
+      x: i9,
+  }
+  ```
+
+  But this proposal is flawed because, in C, both the width and the underlying
+  type influence the layout. The Zig proposal throws the underlying type away.
```

Comment:
This is an extremely small point that I didn't know that affects the proposal a lot. How does it affect it?

This to me is an important aspect that justifies using this method instead of Zig's.

Reply:
I believe the first motivating example should make this clear.

Reply:
Not really. It doesn't do that great a job of providing motivation beyond "this is how it affects C."

Why not just require additional alignment attributes instead of using different types for bitfields? I feel like these should be weighed as additional options.

Compatibility with C definitions could definitely be enough to justify it, but it should be explained better in the RFC rather than assuming every reader knows what it means and how it affects their code.

Reply:
> It doesn't do that great a job of providing motivation beyond "this is how it affects C."

Being compatible with C without having to write target-specific code is, in fact, the motivation of the RFC and this is made pretty clear. Please read the discussion of the original RFC where at least 50% of the over 90 comments were trying to argue that what we actually want is arbitrarily sized integers and custom struct layouts.